### PR TITLE
feat(addon-doc): add type reference handler for improve documentation

### PR DESCRIPTION
**Pull request by splincode**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the new behaviour?

image

image

### CODE REVIEW

File: projects/addon-doc/components/documentation/documentation.module.ts [hunk 1]
```diff
@@ -63,7 +61,7 @@ import {TuiStripOptionalPipe} from './pipes/strip-optional.pipe';
         TuiIsOptionalPipe,
         TuiShowCleanerPipe,
         TuiStripOptionalPipe,
-        TuiShowContentTooltip,
+        TuiDocTypeReferencePipePipe,
```
Comment (waterplea) [inline projects/addon-doc/components/documentation/documentation.module.ts:64]: Duplicated word Pipe

**Review by waterplea — APPROVED** [review]

File: projects/addon-doc/components/documentation/documentation.module.ts (hunk 1 above)
Comment (splincode) [reply projects/addon-doc/components/documentation/documentation.module.ts]: fixed

File: projects/demo/src/modules/app/app.providers.ts [hunk 2]
```diff
@@ -197,6 +198,26 @@ export const APP_PROVIDERS: Provider[] = [
...
             String(tree).replace(baseHref, ``),
     },
+    {
+        provide: TUI_DOC_TYPE_REFERENCE_HANDLER,
+        useValue: (type: string) => {
+            switch (type) {
+                case `any`:
+                case `null`:
+                case `undefined`:
+                case `boolean`:
```
Comment (vladimirpotekhin) [inline projects/demo/src/modules/app/app.providers.ts:208]: nit: add `void`
Снимок экрана 2023-11-28 в 17 20 05

File: projects/demo/src/modules/app/app.providers.ts [hunk 3]
```diff
@@ -197,6 +198,26 @@ export const APP_PROVIDERS: Provider[] = [
...
+                case `string`:
+                case `number`:
+                    return null;
+                case `SafeResourceUrl`:
+                    return `https://angular.dev/api/platform-browser/${type}`;
+                case `PolymorpheusContent`:
```
Comment (vladimirpotekhin) [inline projects/demo/src/modules/app/app.providers.ts:214]: nit: can we handle `T`, `Map<T>`, `PolymorpheusContent<T>` as well? 

`PolymorpheusContent<TuiTreeContext>` refers to an empty search result on this page:

https://taiga-ui--pr6086-splincode-feat-type-6othcr6a.web.app/components/tree/API

**Review by vladimirpotekhin — APPROVED** [review]

File: projects/demo/src/modules/app/app.providers.ts (hunk 2 above)
Comment (splincode) [reply projects/demo/src/modules/app/app.providers.ts:209]: fixed

File: projects/demo/src/modules/app/app.providers.ts (hunk 3 above)
Comment (splincode) [reply projects/demo/src/modules/app/app.providers.ts:218]: fixed

**Review by vladimirpotekhin — APPROVED** [review]

**Review by waterplea — APPROVED** [review]

**Review by waterplea — CHANGES_REQUESTED** [review]
In Button API:
image

I also wonder if we can somehow improve the auto link to the type. Maybe it can search for `type XXX` / `interface XXX`?

**Review by waterplea — APPROVED** [review]
This is great! A few small issues I have managed to find:
image

CustomEvent should probably be handled the same way ReadonlyArray is (see Svg component)

**Comment by splincode** [comment]
@waterplea fixed

**Pull request merged**